clippy_lints/src/manual_let_else.rs

```diff
@@ -0,0 +1,272 @@
+use clippy_utils::diagnostics::span_lint;
+use clippy_utils::higher::IfLetOrMatch;
+use clippy_utils::ty::is_type_diagnostic_item;
+use clippy_utils::visitors::for_each_expr;
+use clippy_utils::{meets_msrv, msrvs, peel_blocks};
+use if_chain::if_chain;
+use rustc_data_structures::fx::FxHashSet;
+use rustc_hir::{Expr, ExprKind, MatchSource, Pat, PatKind, QPath, Stmt, StmtKind};
+use rustc_lint::{LateContext, LateLintPass, LintContext};
+use rustc_middle::lint::in_external_macro;
+use rustc_semver::RustcVersion;
+use rustc_session::{declare_tool_lint, impl_lint_pass};
+use rustc_span::symbol::sym;
+use serde::Deserialize;
+use std::ops::ControlFlow;
+
+declare_clippy_lint! {
+    /// ### What it does
+    ///
+    /// Warn of cases where `let else` could be used
+    ///
+    /// ### Why is this bad?
+    ///
+    /// `let else` provides a standard construct for this pattern
+    /// that people can easily recognize. It's also more compact.
+    ///
+    /// ### Example
+    ///
+    /// ```rust
+    /// # let w = Some(0);
+    /// let v = if let Some(v) = w { v } else { return };
+    /// ```
+    ///
+    /// Could be written:
+    ///
+    /// ```rust
+    /// # #![feature(let_else)]
+    /// # fn main () {
+    /// # let w = Some(0);
+    /// let Some(v) = w else { return };
+    /// # }
+    /// ```
+    #[clippy::version = "1.60.0"]
+    pub MANUAL_LET_ELSE,
+    pedantic,
+    "manual implementation of a let else statement"
+}
+
+pub struct ManualLetElse {
+    msrv: Option<RustcVersion>,
+    matches_behaviour: MatchLintBehaviour,
+}
+
+impl ManualLetElse {
+    #[must_use]
+    pub fn new(msrv: Option<RustcVersion>, matches_behaviour: MatchLintBehaviour) -> Self {
+        Self {
+            msrv,
+            matches_behaviour,
+        }
+    }
+}
+
+impl_lint_pass!(ManualLetElse => [MANUAL_LET_ELSE]);
+
+impl<'tcx> LateLintPass<'tcx> for ManualLetElse {
+    fn check_stmt(&mut self, cx: &LateContext<'_>, stmt: &'tcx Stmt<'tcx>) {
+        let if_let_or_match = if_chain! {
+            if meets_msrv(self.msrv, msrvs::LET_ELSE);
+            if !in_external_macro(cx.sess(), stmt.span);
+            if let StmtKind::Local(local) = stmt.kind;
+            if let Some(init) = local.init;
```

Comment:
```suggestion
            if let Some(init) = local.init;
            if local.els.is_none();
```

We should ensure the current stmt is not already a let-else. We cannot apply it for 3-branches.

```rust
let P(a) = (if let Q(b) = expr { body } else { else1 }) else { else2 };
```

Reply:
This doesn't trigger the lint now, does it? I think the code that stops let else from being considered for linting is in the `IfLetOrMatch::parse(...)`.

Reply:
> This doesn't trigger the lint now, does it? I think the code that stops let else from being considered for linting is in the `IfLetOrMatch::parse(...)`.

Nope. `IfLetOrMatch::parse` can see through parentheses of the `init` expr of the `let`.
Current result, which is inaccurate:
```diff
diff --git a/tests/ui/manual_let_else.rs b/tests/ui/manual_let_else.rs
index bd0ac69e4..0875001a6 100644
--- a/tests/ui/manual_let_else.rs
+++ b/tests/ui/manual_let_else.rs
@@ -194,4 +194,7 @@ fn not_fire() {
         };
     }
     create_binding_if_some_nf!(v, g());
+
+    // Already a let-else.
+    let Some(a) = (if let Some(b) = Some(Some(())) { b } else { return }) else { return; };
 }
diff --git a/tests/ui/manual_let_else.stderr b/tests/ui/manual_let_else.stderr
index b1061a56b..8ab09ab23 100644
--- a/tests/ui/manual_let_else.stderr
+++ b/tests/ui/manual_let_else.stderr
@@ -191,5 +191,11 @@ LL |     create_binding_if_some!(w, g());
    |
    = note: this error originates in the macro `create_binding_if_some` (in Nightly builds, run with -Z macro-backtrace for more info)
 
-error: aborting due to 13 previous errors
+error: this could be rewritten as `let else`
+  --> $DIR/manual_let_else.rs:199:5
+   |
+LL |     let Some(a) = (if let Some(b) = Some(Some(())) { b } else { return }) else { return; };
+   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: consider writing: `let Some(b) = Some(Some(())) else { return };`
+
+error: aborting due to 14 previous errors

```